### PR TITLE
Issue2

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Reference material__:
- https://github.com/plataformatec/simple_form

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by david-l-weitz**
Closes #2

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class ImagesController < ApplicationController
```
Comment (zldavis) [inline app/controllers/images_controller.rb]: I general when possible, we want to try and keep our commits isolated to distinct working functionality. In this case, I wouldn't suggest committing controllers/etc with empty actions and it might make sense to commit the `simple_form` installation/config changes in their own commit.

File: app/assets/javascripts/images.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (zldavis) [inline app/assets/javascripts/images.js]: When you run the rails generator, it will create a bunch of things for you. But all of these files may or may not be needed at this time. You can remove extra files that are not required (there are a few of them).

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  resources :images
```
Comment (zldavis) [inline config/routes.rb]: trying running `rake routes` and see what this does. You might want to make some changes here.

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -1,4 +1,28 @@
 class ImagesController < ApplicationController
+  def index
```
Comment (zldavis) [inline app/controllers/images_controller.rb]: This action seems out of scope for this issue.

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -1,4 +1,28 @@
...
+    @images = Image.all
+  end
+
   def new
+    @image = Image.new()
```
Comment (zldavis) [inline app/controllers/images_controller.rb]: We prefer to omit the parentheses when calling a method that doesn’t accept any parameters. This may take some getting used to 😄, but it is nice to write less characters

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -1,7 +1,15 @@
 require 'test_helper'
+require 'net/http'
 
 class ImagesControllerTest < ActionDispatch::IntegrationTest
-  # test "the truth" do
-  #   assert true
-  # end
+  def test_form
```
Comment (zldavis) [inline test/controllers/images_controller_test.rb]: We usually like to keep the commits complete with production code and accompanying tests.  The unit tests should not be in their own commit.

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -1,7 +1,15 @@
...
+    get '/images/new'
+    assert_select 'form', 'Image Link'
+  end
+
+  def test_error_shown
+    post '/images', params: { image: { link: 'invalid_link'}}
+    assert_select 'span', 'is not a valid URL'
```
Comment (zldavis) [inline test/controllers/images_controller_test.rb]: this is fine for the time being and works, but what do you think might happen if you add more form fields and validations in the future?

File: config/initializers/simple_form.rb [hunk 8]
```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+#
+# Uncomment this and change the path if necessary to include your own
+# components.
+# See https://github.com/plataformatec/simple_form#custom-components to know
+# more about custom components.
+# Dir[Rails.root.join('lib/components/**/*.rb')].each { |f| require f }
+#
+# Use this setup block to configure all options available in SimpleForm.
+SimpleForm.setup do |config|
```
Comment (zldavis) [inline config/initializers/simple_form.rb]: Since we are using [Bootstrap](https://getbootstrap.com/docs/4.1/components/forms/) as our css framework, it is worth checking out the `simple_form` [documentation](https://github.com/plataformatec/simple_form#bootstrap) around how those two play nicely together.

File: app/controllers/images_controller.rb (hunk 1 above)
Comment (david-l-weitz) [reply app/controllers/images_controller.rb]: ah ok. at appfolio do you squash commits upon merge? I'm used to committing freely and then squashing, which is probably not great 👎

File: app/controllers/images_controller.rb (hunk 4 above)
Comment (david-l-weitz) [reply app/controllers/images_controller.rb]: true, what would you do in this case? leave the functionality in?

Comment (david-l-weitz) [reply app/controllers/images_controller.rb]: rm from this issue

File: db/migrate/20190904213309_create_images.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.2]
+  def change
+    create_table :images do |t|
+      t.string :link
```
Comment (zldavis) [inline db/migrate/20190904213309_create_images.rb]: it would make sense to make this a non-nullable column since that is the intention.  It is a good idea to maintain data integrity at the database level even if we are expected to have model validations. We can discuss more on this topic if you would like.

File: test/controllers/images_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+require 'test_helper'
+require 'net/http'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  def test_form
```
Comment (zldavis) [inline test/controllers/images_controller_test.rb]: it seems as though we are still missing some tests to cover all the actions, for example, the `show` action? Similarly to how you structured your tests for the model, you should do the same for each controller action.
